[s3-data-lake] Adding support for the Dremio (-Nessie) catalog

frifriSF59 opened the pull request:
## What
Adding support for the Dremio catalog (currently the Nessie catalog).
This PR also adds the integration tests around those changes (you will need to pull new secrets)

I moved the logic that will generate the catalog properties in the spec itself. i think this is more logical spot for that logic. We should probably try to keep the `S3DataLakeUtil` as generic as possible. We will need to move the rest of the existing logic in there as well in a separate PR.

## Can this PR be safely reverted and rolled back?
- [X] YES 💚
- [ ] NO ❌

frifriSF59 committed:
Update spec to support Dremio

frifriSF59 committed:
generated spec and table id generator

frifriSF59 committed:
Okay that should be all?

frifriSF59 commented on airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/command/iceberg/parquet/IcebergCatalogSpecifications.kt:
```diff
@@ -363,6 +418,79 @@ data class RestCatalogConfiguration(
+@JsonSchemaTitle("Rest Catalog Configuration")
+@JsonSchemaDescription("Rest-specific configuration details for connecting an Iceberg catalog.")
+data class DremioCatalogConfiguration(
+    @JsonSchemaTitle("Dremio Server URI")
```
Having to copy/paste that stuff is annoying

edgao replied:
do we _need_ the jsonschema annotations on the configuration object? I thought we only needed them on the Specification

frifriSF59 replied:
I did not try to remove it. This is mostly a bunch of c/p. The less I think about the spec / conf, the better I feel. I can try

frifriSF59 commented on airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/command/iceberg/parquet/IcebergCatalogSpecifications.kt:
```diff
@@ -363,6 +418,79 @@ data class RestCatalogConfiguration(
+    }
+
+    object CustomNessieConfigConstants {
+        const val CONF_NESSIE_OAUTH2_SUBJECT_TOKEN = "nessie.authentication.oauth2.token-exchange.subject-token"
```
Somehow those do not exist in the nessie constants. I wonder if this a custom thing from Dremio?

frifriSF59 commented on airbyte-integrations/connectors/destination-s3-data-lake/metadata.yaml:
```diff
@@ -24,6 +24,16 @@ data:
             type: GSM
             alias: airbyte-connector-testing-secret-store
+        - name: SECRET_DESTINATION-S3-DATA-LAKE_SYSTEM_AWS_CONFIG
+          fileName: system_aws_config.json
```
This will be use to replace `glue_aws_assume_role`. I think that the current name makes it sound like it is very specific to the Glue use case. That's not the case. It was just added for it originally

frifriSF59 replied:
Okay. looking at the changes you introduced recently in master @edgao , this may be a wrong assumption

edgao replied:
can we delete the glue_aws_assume_role secret?

frifriSF59 replied:
I guess. I was not sure what impact that could have so I was going to introduce the new one and then delete old one in a different PR but I guess let's just rip the bandage

frifriSF59 committed:
Merge branch 'master' into frifri/dremio-catalog

frifriSF59 committed:
Fix bad import after merge conflictsa

edgao commented on airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/command/iceberg/parquet/IcebergCatalogSpecifications.kt:
```diff
@@ -363,6 +418,79 @@ data class RestCatalogConfiguration(
+           This will ONLY be used if the `Destination Namespace` setting for the connection is set to
+           `Destination-defined` or `Source-defined`"""
+    )
+    val namespace: String?,
```
let's make this required

edgao commented on airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/command/iceberg/parquet/IcebergCatalogSpecifications.kt:
```diff
@@ -363,6 +418,79 @@ data class RestCatalogConfiguration(
+    @get:JsonPropertyDescription(
+        """The Personal Access Token generated through your Dremio server"""
+    )
+    val pat: String?
```
is this intentionally not required?

frifriSF59 replied:
nope

frifriSF59 commented:
/bump-version --changelog "Add support for the Dremio (-Nessie) catalog"

edgao commented on airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/command/iceberg/parquet/IcebergCatalogSpecifications.kt:
```diff
@@ -363,6 +418,79 @@ data class RestCatalogConfiguration(
+        val tokenEndpoint = "$baseUri:9047/oauth/token"
+        val serverUri = "$baseUri:19120/api/v2"
+
+        val catalogProperties = buildMap {
```
nit: prefer `mapOf(CatalogUtil.ICEBERG_CATALOG_TYPE to CatalogUtil.ICEBERG_CATALOG_TYPE_NESSIE, ....)` over buildMap when possible

frifriSF59 replied:
I was actually just using the same pattern that we have in https://github.com/airbytehq/airbyte/blob/5f935fbdbf296370713eef809a1548c4b0b55328/airbyte-integrations/connectors/destination-s3-data-lake/src/main/kotlin/io/airbyte/integrations/destination/s3_data_lake/io/S3DataLakeUtil.kt#L232-L304

Curious what the reason is for one vs the other?

edgao commented on airbyte-integrations/connectors/destination-s3-data-lake/src/test-integration/kotlin/io/airbyte/integrations/destination/s3_data_lake/S3DataLakeWriteTest.kt:
```diff
@@ -177,3 +177,14 @@ class NessieMinioWriteTest :
+        Files.readString(S3DataLakeTestUtil.DREMIO_CONFIG_PATH),
+        S3DataLakeDestinationCleaner(
+            S3DataLakeTestUtil.getCatalog(
+                S3DataLakeTestUtil.parseConfig(S3DataLakeTestUtil.DREMIO_CONFIG_PATH),
```
checking: this just points at our self-hosted instance?

(curious: how does the networking stuff work? i.e. is CI on the right VPN, and did we actually set up HTTPS?)

frifriSF59 replied:
This is pointing at the Dremio cluster so those SHOULD be using internal only ips / should not be accessible from the outside. We are using http

octavia-squidington-iii committed:
chore: bump-version

edgao commented on airbyte-integrations/connectors/destination-s3-data-lake/src/main/kotlin/io/airbyte/integrations/destination/s3_data_lake/io/S3DataLakeUtil.kt:
```diff
@@ -220,6 +221,7 @@ class S3DataLakeUtil(
             is GlueCatalogConfiguration ->
                 buildGlueProperties(config, catalogConfig, icebergCatalogConfig, region)
             is RestCatalogConfiguration -> buildRestProperties(config, catalogConfig, s3Properties)
+            is DremioCatalogConfiguration -> catalogConfig.getCatalogProperties(icebergCatalogConfig)
```
👏 this feels like the right pattern

edgao reviewed (approved):
couple minor comments, otherwise :shipit:

frifriSF59 committed:
changelog fix

frifriSF59 committed:
required keys

frifriSF59 committed:
secret cleanup

frifriSF59 committed:
Delete unused constants

frifriSF59 commented:
/format-fix

octavia-squidington-iii committed:
chore: auto-fix lint and format issues

frifriSF59 committed:
Region

# Conflicts:
#	airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/command/iceberg/parquet/IcebergCatalogSpecifications.kt
#	airbyte-integrations/connectors/destination-s3-data-lake/src/main/kotlin/io/airbyte/integrations/destination/s3_data_lake/io/S3DataLakeUtil.kt

frifriSF59 commented:
/format-fix

frifriSF59 commented:
/format-fix

edgao commented:
/format-fix

octavia-squidington-iii committed:
chore: auto-fix lint and format issues

frifriSF59 committed:
Merge branch 'master' into frifri/dremio-catalog

frifriSF59 committed:
Merge branch 'master' into frifri/dremio-catalog
